File: 1721-swapping-nodes-in-a-linked-list/main.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        
        def get_length(node: ListNode) -> int:
            length = 0
            while node:
                length += 1
                node = node.next
            return length
        

        length = get_length(head)

        if k > length:
            return None
        

        kth_node_from_beginning = head
        for _ in range(k-1):
            kth_node_from_beginning = kth_node_from_beginning.next
        

        kth_node_from_end = head
        for _ in range(length - k):
            kth_node_from_end = kth_node_from_end.next
        
        kth_node_from_beginning.val, kth_node_from_end.val = kth_node_from_end.val, kth_node_from_beginning.val

        return head

File: 1721-swapping-nodes-in-a-linked-list/main.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapNodes(head *ListNode, k int) *ListNode {
    
    length := getLength(head)

    if k > length {
        return &ListNode{}
    }

    kThBeggin := head
    for range(k-1) {
        kThBeggin = kThBeggin.Next
    }

    kThEnd := head
    for range(length - k) {
        kThEnd = kThEnd.Next
    }

    kThBeggin.Val, kThEnd.Val = kThEnd.Val, kThBeggin.Val
    return head
}

func getLength(node *ListNode) int {
        var length int = 0
        for node != nil {
            length ++
            node = node.Next
        }

        return length
    }

